clang/lib/StaticAnalyzer/Checkers/StdVariantChecker.cpp

```diff
@@ -0,0 +1,312 @@
+      }();
+      handleConstructorAndAssignment<VariantHeldTypeMap>(Call, C, thisSVal);
+      return true;
+    }
+    return false;
+  }
+
+private:
+  // The default constructed std::variant must be handled separately
+  // by default the std::variant is going to hold a default constructed instance
+  // of the first type of the possible types
+  void handleDefaultConstructor(const CallEvent &Call,
+                                CheckerContext &C) const {
+
+    const auto *AsConstructorCall = dyn_cast<CXXConstructorCall>(&Call);
+    assert(AsConstructorCall && "A constructor call must be passed!");
+
+    SVal ThisSVal = AsConstructorCall->getCXXThisVal();
+
+    const auto *const ThisMemRegion = ThisSVal.getAsRegion();
+    if (!ThisMemRegion) {
+      return;
+    }
+
+    QualType DefaultType =
+        getNthTemplateTypeArgFromVariant(ThisSVal.getType(C.getASTContext())
+                                             .getTypePtr()
+                                             ->getPointeeType()
+                                             .getTypePtr(),
+                                         0);
+
+    ProgramStateRef State = Call.getState();
+    State = State->set<VariantHeldTypeMap>(ThisMemRegion, DefaultType);
+    C.addTransition(State);
+  }
+
+  bool handleStdGetCall(const CallEvent &Call, CheckerContext &C) const {
+    ProgramStateRef State = Call.getState();
+
+    const auto &ArgType = Call.getArgSVal(0)
+                              .getType(C.getASTContext())
+                              .getTypePtr()
+                              ->getPointeeType()
+                              .getTypePtr();
+    // We have to make sure that the argument is an std::variant.
+    // There is another std::get with std::pair argument
+    if (!isStdVariant(ArgType)) {
+      return false;
+    }
+
+    // Get the mem region of the argument std::variant and get what type
+    // information is known about it.
+    const MemRegion *ArgMemRegion = Call.getArgSVal(0).getAsRegion();
+    const QualType *StoredType = State->get<VariantHeldTypeMap>(ArgMemRegion);
+    if (!StoredType) {
+      return false;
+    }
+
+    const auto &TypeOut = getFirstTemplateArgument(Call);
+    // std::get's first template parameter can be the type we want to get
+    // out of the std::variant or a natural number which is the position of
+    // the wished type in the argument std::variant's type list.
+    auto RetrievedType = [&]() {
+      switch (TypeOut.getKind()) {
+      case TemplateArgument::ArgKind::Type:
+        return TypeOut.getAsType();
+      case TemplateArgument::ArgKind::Integral:
+        // In the natural number case we look up which type corresponds to the
+        // number.
+        return getNthTemplateTypeArgFromVariant(
+            ArgType, TypeOut.getAsIntegral().getSExtValue());
+      default:
+        llvm_unreachable("An std::get's first template argument can only be a "
+                         "type or an integral");
```

Comment:
`llvm_unreachable` should be only really rarely used. Only where it's mandated and justified.
Letting the compiler know that it could optimize assuming it's unreachable it a dangerous territory.